Feature/parse file by path

eric-ozim opened the pull request:
## PR Description

### Feature: File Parse by Path with S3 Support

This PR implements a new API endpoint that enables file parsing via file path (local or S3 URI), with optimized response strategies for different client needs.

---

## What Changed

### New API Endpoint
- **`POST /file_parse_by_path`** - New endpoint for parsing files by path (local or S3)

### New Functions
- **`read_file_with_s3_support()`** - Unified file reader supporting both local paths and S3 URIs (68 lines)

### Modified Files
- **`mineru/cli/fast_api.py`** - Added new endpoint and helper functions

**Total Changes:** +253 lines, -2 lines

---

## Key Features

### 1. Unified File Access

**Supports Multiple File Sources:**
- **Local Path**: `/data/docs/report.pdf`
- **S3 URI**: `s3://bucket-name/path/to/file.pdf`

**S3 Configuration Options:**
- **Option 1 (Recommended)**: Environment variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `S3_ENDPOINT_URL`)
- **Option 2**: API parameters (`s3_ak`, `s3_sk`, `s3_endpoint`)

### 2. Optimized Response Strategy

**Lightweight Mode (default, `return_content=False`):**
- Returns only file paths and sizes
- Minimizes network overhead
- Client reads files locally

**Full Response Mode (`return_content=True`):**
- Returns file contents embedded in response
- Suitable for remote clients without file system access

### 3. Smart Output Directory Handling

- **Local Files**: Auto-generates output in source file directory if not specified
- **S3 Files**: Requires explicit `output_dir` parameter

### 4. Image-to-PDF Conversion

- Automatically converts image files (JPG, PNG, etc.) to PDF format
- Supports both local and S3 image sources

---

## Implementation Details

### File Reading Logic

```python
read_file_with_s3_support(file_path, s3_ak, s3_sk, s3_endpoint, s3_addressing_style)
```

1. **Path Detection**: Checks if path starts with `s3://` or `s3a://`
2. **S3 Handling**: Uses `S3Reader` to fetch file bytes from object storage
3. **Local Handling**: Uses existing `read_fn()` for local file system
4. **Format Detection**: Auto-detects file type via `guess_suffix_by_bytes()`
5. **Image Conversion**: Converts images to PDF using `images_bytes_to_pdf_bytes()`

### Response Structure

**Lightweight Response (default):**
```json
{
  "results": {
    "filename": {
      "output_dir": "/path/to/output",
      "parse_dir": "/path/to/output/filename/auto",
      "markdown_path": "/path/to/output/filename/auto/filename.md",
      "markdown_size": 12345,
      "images_dir": "/path/to/output/filename/auto/images",
      "images_count": 5
    }
  }
}
```

**Full Response (`return_content=True`):**
```json
{
  "results": {
    "filename": {
      "...paths and sizes...",
      "content": {
        "md_content": "...",
        "middle_json": {...},
        "images": {"image1.jpg": "data:image/jpeg;base64,..."}
      }
    }
  }
}
```

---

## New Dependencies

- `mineru.data.io.s3.S3Reader` - S3 file reader
- `mineru.data.utils.path_utils.parse_s3path` - S3 path parser
- `mineru.utils.guess_suffix_or_lang.guess_suffix_by_bytes` - Byte-based format detection
- `mineru.utils.pdf_image_tools.images_bytes_to_pdf_bytes` - Image-to-PDF converter

---

## API Usage Examples

### Example 1: Parse Local File (Auto Output Dir)
```bash
curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=/data/docs/report.pdf" \
  -F "lang=en"
```

### Example 2: Parse S3 File (With Credentials)
```bash
curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=s3://my-bucket/docs/report.pdf" \
  -F "output_dir=/tmp/mineru_output" \
  -F "s3_ak=YOUR_ACCESS_KEY" \
  -F "s3_sk=YOUR_SECRET_KEY" \
  -F "s3_endpoint=https://s3.amazonaws.com"
```

### Example 3: Parse S3 File (Using Env Variables)
```bash
export AWS_ACCESS_KEY_ID=your_key
export AWS_SECRET_ACCESS_KEY=your_secret
export S3_ENDPOINT_URL=https://s3.amazonaws.com

curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=s3://my-bucket/docs/report.pdf" \
  -F "output_dir=/tmp/mineru_output"
```

### Example 4: Get Full Response with Content
```bash
curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=/data/docs/report.pdf" \
  -F "return_content=true"
```

---

## Jira

-https://implustech.atlassian.net/browse/OZIM-729

---

## ✅ Checklist

### Code Quality
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my own code
- [] I have commented my code, particularly in hard-to-understand areas
- [] My changes generate no new warnings

### Testing
- [] I have tested local file parsing
- [] I have tested S3 file parsing with credentials
- [] I have tested S3 file parsing with environment variables
- [] I have tested both response modes (lightweight and full)


---

## 🧪 Testing Guide

### Test Scenario 1: Local File Parse
```bash
# Prepare test file
echo "test content" > /tmp/test.txt

# Call API
curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=/tmp/test.txt"

# Verify: Check response contains output_dir and parse_dir
```

### Test Scenario 2: S3 File Parse
```bash
# Set S3 credentials
export AWS_ACCESS_KEY_ID=your_key
export AWS_SECRET_ACCESS_KEY=your_secret
export S3_ENDPOINT_URL=https://s3.amazonaws.com

# Call API
curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=s3://test-bucket/sample.pdf" \
  -F "output_dir=/tmp/s3_output"

# Verify: Check file is downloaded and parsed correctly
```

### Test Scenario 3: Response Mode Comparison
```bash
# Lightweight mode (default)
time curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=/tmp/test.pdf" \
  -F "return_content=false"

# Full mode (with content)
time curl -X POST "http://localhost:8080/file_parse_by_path" \
  -F "file_path=/tmp/test.pdf" \
  -F "return_content=true"

# Verify: Full mode takes longer but returns content in response
```

### Expected Logs:

eric-ozim committed:
api(feat): add file_parse_by_path endpoint for parsing files from server local path

- Add new POST endpoint /file_parse_by_path for parsing files without upload
- Support flexible output directory configuration with fallback to default
- Add writable directory validation and auto-fallback mechanism
- Support selective file generation (markdown, json, images, etc.)
- Provide lightweight response with metadata only by default
- Optional content inclusion in response via return_content parameter
- Remove unused asyncio import

eric-ozim committed:
api(refactor): optimize file_parse_by_path response logic

- Use loop to unify file path and size return logic
- Eliminate code duplication for better maintainability

eric-ozim committed:
api(feat): add S3 support to parse_pdf_by_path endpoint

myhloli commented:
Thank you for submitting your code to the repository!
However, I have a few questions :  
The API service is designed to decouple the devices submitting tasks from those processing (parsing) tasks via a client-server architecture. Why should local file parsing and returning of local paths be supported within the API?

eric-ozim committed:
api(feat): implement S3 output upload and dynamic response format

- Add upload_directory_to_s3() to write parse results to S3
- Add get_s3_config() to load credentials from environment variables
- Update response format: use '_uri' suffix for S3, '_path' for local
- Add 'storage_type' field to indicate output storage type
- Fix content reading order: read before S3 upload and temp cleanup
- Support cross-storage operations (S3->local, local->S3)
- Environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, S3_ENDPOINT_URL

eric-ozim committed:
api(refactor): unify path fields in response

- Use *_path for both S3 and local paths
- Remove redundant _uri/_path separation
- Determine storage from storage_type or s3:// prefix
